pkg-manager/resolve-dependencies/src/resolveDependencies.ts

```diff
@@ -830,6 +823,44 @@ async function resolveDependenciesOfDependency (
     supportedArchitectures: options.supportedArchitectures,
     parentIds: options.parentIds,
   }
+
+  // The catalog protocol is normally replaced when resolving the dependencies
+  // of importers. However, when a workspace package is "injected", it becomes a
+  // "file:" dependency and is no longer an "importer" from the perspective of
+  // pnpm.
+  //
+  // To allow the catalog protocol to still be used for injected workspace
+  // packages, it's necessary to check if the parent package was an injected
+  // workspace package and replace the catalog: protocol for the current package.
+  const isInjectedWorkspacePackage = options.parentPkg.resolvedVia === 'workspace' &&
+    options.parentPkg.pkgId.startsWith('file:')
```

Comment:
I don't think the second check is required.
```suggestion
  const isInjectedWorkspacePackage = options.parentPkg.resolvedVia === 'workspace'
```

Reply:
It's not required if everything works as expected, but I wanted to be defensive against a bug in the future with more explicit checks.

The intention was to only replace `catalog:` protocol references of injected workspace packages, and I wanted the `file:` check to help make sure we're only replacing the `catalog:` protocol in that specific case.

If the normal process of replacing the `catalog:` protocol in importers doesn't work due to a new bug, the code here (if it didn't check for `file:`) would end up replacing the `catalog:` protocol. I was worried that would make tracking down unexpected situations harder since they would fail silently.

Reply:
Can this single place used for importers too?

Reply:
We could with more changes. At the moment the `resolvedVia` field isn't populated when an importer is the `options.parentPkg`. (It's populated when it's an importer depending on another importer.) I think we would need a new field here:

https://github.com/pnpm/pnpm/blob/5f7be645e4c0754bb98dec7b76fee4e00ac5da50/pkg-manager/resolve-dependencies/src/resolveDependencyTree.ts#L199-L207

I do see the existing `rootDir` field. If that's only set for importers we could reuse that field, but it might be clearer to do a new `isWorkspacePackage` field to avoid bugs if `rootDir` is set for other packages in the future.

---

Separately, I did like how we spent time refactoring the importer code path to make the `catalog:` protocol only replaced in `resolveDependenciesOfImporterDependency` and not deeper down the dependency resolution call stack. I would be okay regressing that, but I was hoping to preserve that setup for the fix here.